### DISCUSSION TITLE
Uni-math is here!

### DISCUSSION THREAD

**Discussion by NanoMichael** [NanoMichael/MicroTeX#87] [Announcements]
After a few months of coding (and struggling), finally, the Uni-math is (almost) here! :fireworks:

First, let me show you some screenshots to give an intuitive feel for this major update:

- emoji support

![Screenshot from 2021-09-12 02-02-04](https://user-images.githubusercontent.com/20122696/132957061-834ed63b-7d43-4970-8ee9-0c6dde10a2c9.png)

- multi-fonts support

![Screenshot from 2021-09-12 02-10-16](https://user-images.githubusercontent.com/20122696/132957256-ec2a5d16-0742-4146-87b5-1bbcb0123720.png)

# Major changes

- Uni-math support, unlike the Plain-Old-TeX, now we just need one math font to display formulas, so the font management is really simple
- better architecture, `tinyxml2` is no longer needed
- no `std::wstring` anymore
- support for 2877 math symbols (include alias)
- multi-fonts support
- improved debugging info

And much more I haven't listed here...

# Play with the GTK demo application

## Build the demo

Since I have not yet provided an executable to demonstrate this program, you can build the demo by yourself.

First, make sure you have `gtkmm-3.0` and `gtksourceviewmm-3.0` installed, the GTK sample application will use it to show a simple GUI. Then use `cmake`  to build it:

```shell
mkdir build && cd build && cmake .. && make -j8
```

also, you can build it with meson:

```shell
meson setup builddir && cd builddir && meson compile
```

## Generate clm data file

The `clm` data file stores the font metrics and other OTF font information to layout math formulas. This program provides a script located in folder `prebuilt` to generate a `clm` data file from an OTF font. First, make sure you have `fontforge` installed, then run the following command:

```shell
fontforge -lang=py -script otf2clm.py \
  path/to/math/font \ #1
  true \ #2
  path/to/save/clm #3
```

- argument 1: specifies the file path of the OTF font to be parsed
- argument 2: specifies if the OTF font is a math font
- argument 3: specifies the path to save the generated `clm` data file

You can download `XITS` from [here](https://ctan.org/pkg/xits), and `latin-modern math` from [here](https://www.ctan.org/pkg/lm-math).

## Run the demo

After all works done, you can run the demo application now:

```shell
./LaTeX \
  -mathfont=path/to/math/font \ #1
  -clm=path/to/clm/data/file \ #2
  -samples=path/to/samples/file #3
```

- argument 1: specifies the math font file to display formulas
- argument 2: specifies the `clm` data file to layout the formulas
- argument 3: specifies the file containing various samples, you can use `./res/SAMPLES.tex`

## Switch font

You can add math font via:

```tex
\addmathfont{% font name
  xits
}{% font file path
  font/file/path
}{% clm data file path
  clm/data/file/path
}
```

And switch font via:

```tex
\mathversion{xits}
% formulas below will use the xits font
```

# And the logo

I've designed the logo shown below, but I'm not satisfied with it... Any suggestions?

![Screenshot from 2021-09-13 00-31-36](https://user-images.githubusercontent.com/20122696/132995452-fdde284c-82e8-433d-906c-566b7fe84980.png)

**Comment by sp1ritCS** [comment]
have you seen the comment I made on commit 31283e1?

https://github.com/NanoMichael/cLaTeXMath/commit/31283e1532bac77249ea75f724246294b062433c#r56210140

**Reply by NanoMichael** [reply]
done by commit ef33620 :fireworks:

**Comment by sp1ritCS** [comment]
thanks <3

could you also bump the version from 0.0.4 to 0.1.0 (or maybe even 1.0.0), given that this contains fairly major changes?

**Reply by NanoMichael** [reply]
I'm planning to bump the version from 0.0.4 to 1.0.0 (since it's a huge update), but some works need to be done. First let me finish this announcement :smile:

**Comment by sp1ritCS** [comment]
So, after playing with it for a while, I've noticed a few things.

## documentation & invoke params
I assume this is something you are aware of, but the parser you use for launch parameter fails to parse is incorrect (value ends up being the key) cause it to throw tex:ex_file_not_found (or similar). Consider using gnu getopts instead.

Additionally a bit of documentation would be useful to understand what `clm` and `mathfont` is.

## displaystyle
clm used to default to displaystyle for equations. it doesn't do so anymore
In order to get that behaviour in LaTeX I'd have to add `\everymath{\displaystyle}`. However that isn't possible with clm  because it doesn't support `\everymath`.

also clm segfaults when nothing follows `\displaystyle`. (this behaviour is also present in current clm)

## fallback fonts
Do you intend to ship a fallback font with  clm in case an application just wants to render latex without the need to manage a custom font?

**Reply by NanoMichael** [reply]
> documentation & invoke params

The options handling is really rough since it is just a demo application :sweat_smile:.  I'll optimize it in the future, and the documentation is on the way.

> displaystyle

I'll add a function to config the default style. I can't reproduce the `segfaults` with `\displaystyle`, could you provide an example?

> fallback fonts

Probably no. In my submission, it is the user's responsibility to provide the font with the `clm` data file, but it needs to be discussed.

**Reply by sp1ritCS** [reply]
> I'll add a function to config the default style. I can't reproduce the segfaults with \displaystyle, could you provide an example?

Ah, I see this only occurs when having it wrapped in $. Due to how [NoteKit](https://github.com/blackhole89/notekit)'s parser for mardown LaTeX snippets work, these are always included.

Giving the demo application `$\displaystyle$` it'll segfault too

**Reply by NanoMichael** [reply]
OK... I probably get the cause now... I'll fix it later

**Reply by NanoMichael** [reply]
`\everymath` now supported via commit 01b3ca6, and the `segfaults` when parsing `$\displaystyle$` is fixed by commit cc28e8d

**Reply by NanoMichael** [reply]
Ummm, I've forgotten the return statement... fixed by 5a418e3

**Reply by sp1ritCS** [reply]
yeah, I just wanted to test it but it segfaulted ._.

Edit: but now it works 🚀

**Comment by LyesSaadi** [comment]
OMG, I'm in awe of the first image there, so amazing yet so cursed.

On another note, good job on doing that O.o ! That should have been such an immense amount of work :clap: :tada: !
